beps/0006-scaffolder-action-rollback/README.md

```diff
@@ -1,84 +1,303 @@
+      reason: string;
+    };
+```
+
+This enhancement allows us to identify checkpoints that failed to rollback for potential manual intervention, thus tracking the complete lifecycle of a checkpoint (creation → potential rollback).
+
+### WorkflowRunner Enhancements
+
+The `NunjucksWorkflowRunner` will be modified to track successfully completed checkpoints with their rollback functions. When a scaffolder task fails, the system will:
+
+1. Query the database for completed checkpoints associated with the task
+2. Filter for checkpoints that have rollback functions defined
+3. Execute these rollback functions in reverse order
+
+The implementation will ensure that:
+
+- Rollbacks are executed in proper LIFO order across all actions
+- Each rollback operation is properly logged and tracked
+- **Failures in one rollback don't prevent other rollbacks from being attempted**
+
+```typescript
+// Pseudo-code for rollback execution
+for (const checkpoint of completedCheckpointsWithRollback.reverse()) {
+  try {
+    await checkpoint.rollback(checkpoint.data);
+    logger.info(`Successfully rolled back checkpoint: ${checkpoint.key}`);
+  } catch (error) {
+    logger.error(`Failed to rollback checkpoint ${checkpoint.key}: ${error}`);
+    // Continue with next rollback
+  }
+}
+```
+
+### Enabling Rollback for Checkpoints
+
+!!! warning Work in progress
+
+Rollbacks will be disabled by default and need to be explicitly enabled. We will add a flag to enable rollback functionality at the task level, with the option to override this at the action level.
+
+```yaml
+apiVersion: scaffolder.backstage.io/v1beta3
+kind: Template
+metadata:
+  name: example-template
+spec:
+  parameters:
+    # ...
+  steps:
+    - id: github-repo
+      name: Create GitHub Repository
+      action: github:repo:create
+      rollbackEnabled: true # Enable rollback for this action
+      input:
+        # ...
+```
+
+Additionally, we will add a system-level configuration option to enable/disable rollback functionality globally.
+
+### Auditing and Permissions
+
+1. Implement the [AuditorService](https://backstage.io/docs/backend-system/core-services/auditor) for key operations related to checkpoints and rollbacks
+2. Add a rollback permission definition with permission checks in the `NunjucksWorkflowRunner`
+3. Add specific rules for rollback operations to `scaffolderActionRules`
+4. Log all rollback attempts and outcomes to provide a clear audit trail
+
+### Securing Rollbacks
+
+!!! warning Is this still needed?
+
+For checkpoints that require additional security checks during rollback, we provide a secure rollback helper function. This function will ensure that the resources being rolled back match what was originally created by validating the checkpoint data.
 
 ```typescript
-const createPublishGitHubAction = createTemplateAction({
-  id: 'publish:github',
-  async handler() {},
-  async rollback() {},
-});
+export function createSecureCheckpointRollback<T>(
```

Comment:
I'm wondering if we should be encouraging resources are decorated with the `taskId` or something like that.

Take for example creating a `GitHub` repository, I know you can add attributes to the repository, and wondering if we should be adding something like `backstage.io/scaffolder-resource: ${taskId}` or something to all of the things it creates so that we can do the checking in the rollback function to make sure that these things were created by the scaffolder.

I'm thinking that if we implement it like the below as a higher order function, then it's not a strong requirement to tackle this straight away.